File: XTApp/XTApp/XTCommon/XTCategory/UIView+XTCategory.m
//
//  UIView+XTCategory.m
//  XTApp
//
//  Created by xia on 2024/7/12.
//

#import "UIView+XTCategory.h"

@implementation UIView (XTCategory)

/**
 *  Custom initialization method
 *
 *  @param frame         frame
 *  @param color  颜色
 *
 */
+(UIView *)xt_frame:(CGRect)frame
              color:(UIColor * __nullable)color {
    UIView *view = [[UIView alloc] initWithFrame:frame];
    view.backgroundColor = color;
    return view;
}

/**
 *  渐变图层
 *
 *  @param colors         颜色数组
 *  @param startPoint 起始点
 *  @param endPoint 结束点
 *  @param size 大小
 *
 */
+(CAGradientLayer *)xt_layer:(NSArray <id>*)colors
                   locations:(NSArray *)locations
                  startPoint:(CGPoint)startPoint
                    endPoint:(CGPoint)endPoint
                        size:(CGSize)size {
    CAGradientLayer *layer = [CAGradientLayer layer];
    layer.colors = colors;
    layer.locations = locations;
    layer.startPoint = startPoint;
    layer.endPoint = endPoint;
    layer.frame = CGRectMake(0, 0, size.width, size.height);
    return layer;
}

/**
 *  给view加圆角
 *
 *  @param rect 大小
 *  @param corners 圆角类型
 *  @param size 圆角大小
 *
 */
-(void)xt_rect:(CGRect)rect
       corners:(UIRectCorner)corners
          size:(CGSize)size {
    UIBezierPath *path = [UIBezierPath bezierPathWithRoundedRect:rect byRoundingCorners:corners  cornerRadii:size];
    CAShapeLayer *layer = [[CAShapeLayer alloc]init];
    //设置大小
    layer.frame = rect;
    //设置图形样子
    layer.path = path.CGPath;
    self.layer.mask = layer;
}

+ (UIImageView *)xt_img:(NSString *)name
                    tag:(NSInteger)tag{
    UIImageView *img = [UIImageView new];
    [img setContentCompressionResistancePriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisHorizontal];
    [img setContentHuggingPriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisHorizontal];
    
    [img setContentCompressionResistancePriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisVertical];
    [img setContentHuggingPriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisVertical];
    img.tag = tag;
    if([name isKindOfClass:[NSString class]] && name.length > 0)
    {
        [img setImage:[UIImage imageNamed:name]];
    }
    
    return img;
}

+ (UIButton *)xt_btn:(NSString *)title
                font:(UIFont *)font
           textColor:(UIColor *)textColor
        cornerRadius:(float)cornerRadius
         borderColor:(UIColor *)borderColor
         borderWidth:(float)borderWidth
     backgroundColor:(UIColor *)backgroundColor
                 tag:(NSInteger)tag {
    UIButton *btn = [UIButton new];
    [btn setBackgroundColor:[UIColor clearColor]];
    if (title)
        [btn setTitle:title forState:UIControlStateNormal];
    if (font)
        [btn.titleLabel setFont:font];
    if (textColor)
        [btn setTitleColor:textColor forState:UIControlStateNormal];
    if (cornerRadius > 0)
        btn.layer.cornerRadius = cornerRadius;
    if (borderColor)
        btn.layer.borderColor = borderColor.CGColor;
    if (borderWidth > 0)
        btn.layer.borderWidth = borderWidth;
    if (backgroundColor)
        btn.backgroundColor = backgroundColor;
    [btn setTag:tag];
    return btn;
}

+(UIButton *)xt_btn:(NSString * __nullable)title
               font:(UIFont * __nullable)font
          textColor:(UIColor * __nullable)textColor
       cornerRadius:(float)cornerRadius
                tag:(NSInteger)tag {
    UIButton *btn = [UIButton new];
    [btn setBackgroundColor:[UIColor clearColor]];
    if (title)
        [btn setTitle:title forState:UIControlStateNormal];
    if (font)
        [btn.titleLabel setFont:font];
    if (textColor)
        [btn setTitleColor:textColor forState:UIControlStateNormal];
    if (cornerRadius > 0) {
        btn.layer.cornerRadius = cornerRadius;
        btn.clipsToBounds = YES;
    }
    [btn setTag:tag];
    return btn;
}

+(UITextField *)xt_textField:(BOOL)secureText
                 placeholder:(NSString *)placeholder
                        font:(UIFont *)font
                   textColor:(UIColor *)textColor
                withdelegate:(id)delegate {
    UITextField *textField = [UITextField new];
    
    textField.clearButtonMode = UITextFieldViewModeWhileEditing;
    if (secureText)
        textField.secureTextEntry = YES;
    if (placeholder) {
        [textField xt_placeholder:placeholder placeholderColor:XT_RGB(0xB0B0B0, 1.0f)];
    }
    if(textColor) {
        textField.textColor = textColor;
    }
    textField.font = font;
    
    if(delegate)
    {
        textField.delegate = delegate;
    }
    return textField;
}

+ (UILabel *)xt_lab:(CGRect)frame
               text:(NSString *)text
               font:(UIFont *)font
          textColor:(UIColor *)textColor
          alignment:(NSTextAlignment)alignment
                tag:(NSInteger)tag {
    UILabel *lab = [[UILabel alloc] initWithFrame:frame];
    [lab setText:text];
    [lab setTextColor:textColor];
    [lab setFont:font];
    lab.textAlignment = alignment;
    [lab setBackgroundColor:[UIColor clearColor]];
    [lab setTag:tag];
    return lab;
}

+ (UILabel *)xt_lab:(NSString *)text
               font:(UIFont *)font
          textColor:(UIColor *)textColor
          alignment:(NSTextAlignment)alignment
         isPriority:(BOOL)isPriority
                tag:(NSInteger)tag {
    UILabel *lab = [UILabel xt_lab:CGRectZero text:text font:font textColor:textColor alignment:alignment tag:tag];
    
    if(isPriority){
        [lab setContentCompressionResistancePriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisHorizontal];
        [lab setContentHuggingPriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisHorizontal];
        
        [lab setContentCompressionResistancePriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisVertical];
        [lab setContentHuggingPriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisVertical];
    }
    return lab;
}

//画虚线 - draw dash line.
- (void)xt_drawLineFromPoint:(CGPoint)fPoint
                     toPoint:(CGPoint)tPoint
                   lineColor:(UIColor *)color
                   lineWidth:(CGFloat)width
                  lineHeight:(CGFloat)height
                   lineSpace:(CGFloat)space
                    lineType:(NSInteger)type {
    CAShapeLayer* shapeLayer = [CAShapeLayer layer];
    shapeLayer.strokeColor = [UIColor lightGrayColor].CGColor;
    if (color) {
        shapeLayer.strokeColor = color.CGColor;
    }
    shapeLayer.fillColor = [UIColor clearColor].CGColor;
    shapeLayer.path = ({
        UIBezierPath *path = [UIBezierPath bezierPath];
        [path moveToPoint:fPoint];
        [path addLineToPoint:tPoint];
        path.CGPath;
    });
    //第一格虚线缩进多少 - the degree of indent of the first cell
    //shapeLayer.lineDashPhase = 4;
    shapeLayer.lineWidth = height < 0 ? 1 : height;
    shapeLayer.lineCap = kCALineCapButt;
    width = width < 0 ? 1 : width;
    shapeLayer.lineDashPattern = @[@(width),@(space)];
    if (type == 1) {
        shapeLayer.lineCap = kCALineCapRound;
        shapeLayer.lineDashPattern = @[@(width),@(space+width)];
    }
    [self.layer addSublayer:shapeLayer];
}

@end

@implementation UITextField (XTCategory)

- (void)xt_placeholder:(NSString *)placeholder
      placeholderColor:(UIColor *)placeholderColor {
    self.attributedPlaceholder = [[NSAttributedString alloc] initWithString:placeholder attributes:@{NSForegroundColorAttributeName:placeholderColor}];
}

@end

@implementation UIViewController (XTCategory)

- (void)xt_presentViewController:(UIViewController *)viewControllerToPresent
                        animated: (BOOL)flag
                      completion:(void (^ __nullable)(void))completion
          modalPresentationStyle:(UIModalPresentationStyle)modalPresentationStyle {
    viewControllerToPresent.modalPresentationStyle = modalPresentationStyle;
    [self presentViewController:viewControllerToPresent animated:flag completion:completion];
}

@end

File: XTApp/XTApp/XTCommon/XTCategory/UIView+XTCategory.swift
//
//  UIView+XTCategory.swift
//  XTApp
//
//  Created by Codex on 2026/4/26.
//

import UIKit

private func xtCategoryColor(_ rgbValue: Int, alpha: CGFloat = 1.0) -> UIColor {
    UIColor(
        red: CGFloat((rgbValue & 0xFF0000) >> 16) / 255.0,
        green: CGFloat((rgbValue & 0x00FF00) >> 8) / 255.0,
        blue: CGFloat(rgbValue & 0x0000FF) / 255.0,
        alpha: alpha
    )
}

@objc
extension UIView {
    @objc(xt_frame:color:)
    class func xt_frame(_ frame: CGRect, color: UIColor?) -> UIView {
        let view = UIView(frame: frame)
        view.backgroundColor = color
        return view
    }

    @objc(xt_layer:locations:startPoint:endPoint:size:)
    class func xt_layer(_ colors: [Any], locations: [Any], startPoint: CGPoint, endPoint: CGPoint, size: CGSize) -> CAGradientLayer {
        let layer = CAGradientLayer()
        layer.colors = colors
        layer.locations = locations as? [NSNumber]
        layer.startPoint = startPoint
        layer.endPoint = endPoint
        layer.frame = CGRect(x: 0, y: 0, width: size.width, height: size.height)
        return layer
    }

    @objc(xt_rect:corners:size:)
    func xt_rect(_ rect: CGRect, corners: UIRectCorner, size: CGSize) {
        let path = UIBezierPath(roundedRect: rect, byRoundingCorners: corners, cornerRadii: size)
        let maskLayer = CAShapeLayer()
        maskLayer.frame = rect
        maskLayer.path = path.cgPath
        layer.mask = maskLayer
    }

    @objc(xt_img:tag:)
    class func xt_img(_ name: String?, tag: Int) -> UIImageView {
        let imageView = UIImageView()
        imageView.setContentCompressionResistancePriority(.required, for: .horizontal)
        imageView.setContentHuggingPriority(.required, for: .horizontal)
        imageView.setContentCompressionResistancePriority(.required, for: .vertical)
        imageView.setContentHuggingPriority(.required, for: .vertical)
        imageView.tag = tag
        if let name, !name.isEmpty {
            imageView.image = UIImage(named: name)
        }
        return imageView
    }

    @objc(xt_btn:font:textColor:cornerRadius:borderColor:borderWidth:backgroundColor:tag:)
    class func xt_btn(
        _ title: String?,
        font: UIFont?,
        textColor: UIColor?,
        cornerRadius: Float,
        borderColor: UIColor?,
        borderWidth: Float,
        backgroundColor: UIColor?,
        tag: Int
    ) -> UIButton {
        let button = UIButton(type: .custom)
        button.backgroundColor = .clear
        if let title {
            button.setTitle(title, for: .normal)
        }
        if let font {
            button.titleLabel?.font = font
        }
        if let textColor {
            button.setTitleColor(textColor, for: .normal)
        }
        if cornerRadius > 0 {
            button.layer.cornerRadius = CGFloat(cornerRadius)
        }
        if let borderColor {
            button.layer.borderColor = borderColor.cgColor
        }
        if borderWidth > 0 {
            button.layer.borderWidth = CGFloat(borderWidth)
        }
        if let backgroundColor {
            button.backgroundColor = backgroundColor
        }
        button.tag = tag
        return button
    }

    @objc(xt_btn:font:textColor:cornerRadius:tag:)
    class func xt_btn(
        _ title: String?,
        font: UIFont?,
        textColor: UIColor?,
        cornerRadius: Float,
        tag: Int
    ) -> UIButton {
        let button = UIButton(type: .custom)
        button.backgroundColor = .clear
        if let title {
            button.setTitle(title, for: .normal)
        }
        if let font {
            button.titleLabel?.font = font
        }
        if let textColor {
            button.setTitleColor(textColor, for: .normal)
        }
        if cornerRadius > 0 {
            button.layer.cornerRadius = CGFloat(cornerRadius)
            button.clipsToBounds = true
        }
        button.tag = tag
        return button
    }

    @objc(xt_textField:placeholder:font:textColor:withdelegate:)
    class func xt_textField(
        _ secureText: Bool,
        placeholder: String?,
        font: UIFont?,
        textColor: UIColor?,
        withdelegate delegate: Any?
    ) -> UITextField {
        let textField = UITextField()
        textField.clearButtonMode = .whileEditing
        if secureText {
            textField.isSecureTextEntry = true
        }
        if let placeholder {
            textField.xt_placeholder(placeholder, placeholderColor: xtCategoryColor(0xB0B0B0))
        }
        if let textColor {
            textField.textColor = textColor
        }
        textField.font = font
        if let delegate = delegate as? UITextFieldDelegate {
            textField.delegate = delegate
        }
        return textField
    }

    @objc(xt_lab:text:font:textColor:alignment:tag:)
    class func xt_lab(_ frame: CGRect, text: String?, font: UIFont?, textColor: UIColor?, alignment: NSTextAlignment, tag: Int) -> UILabel {
        let label = UILabel(frame: frame)
        label.text = text
        label.textColor = textColor
        label.font = font
        label.textAlignment = alignment
        label.backgroundColor = .clear
        label.tag = tag
        return label
    }

    @objc(xt_lab:font:textColor:alignment:isPriority:tag:)
    class func xt_lab(_ text: String?, font: UIFont?, textColor: UIColor?, alignment: NSTextAlignment, isPriority: Bool, tag: Int) -> UILabel {
        let label = xt_lab(.zero, text: text, font: font, textColor: textColor, alignment: alignment, tag: tag)
        if isPriority {
            label.setContentCompressionResistancePriority(.required, for: .horizontal)
            label.setContentHuggingPriority(.required, for: .horizontal)
            label.setContentCompressionResistancePriority(.required, for: .vertical)
            label.setContentHuggingPriority(.required, for: .vertical)
        }
        return label
    }

    @objc(xt_drawLineFromPoint:toPoint:lineColor:lineWidth:lineHeight:lineSpace:lineType:)
    func xt_drawLine(from fPoint: CGPoint, to tPoint: CGPoint, lineColor color: UIColor?, lineWidth width: CGFloat, lineHeight height: CGFloat, lineSpace space: CGFloat, lineType type: Int) {
        let shapeLayer = CAShapeLayer()
        shapeLayer.strokeColor = (color ?? .lightGray).cgColor
        shapeLayer.fillColor = UIColor.clear.cgColor
        let path = UIBezierPath()
        path.move(to: fPoint)
        path.addLine(to: tPoint)
        shapeLayer.path = path.cgPath
        let lineWidth = width < 0 ? 1 : width
        shapeLayer.lineWidth = height < 0 ? 1 : height
        shapeLayer.lineCap = .butt
        shapeLayer.lineDashPattern = [NSNumber(value: Double(lineWidth)), NSNumber(value: Double(space))]
        if type == 1 {
            shapeLayer.lineCap = .round
            shapeLayer.lineDashPattern = [NSNumber(value: Double(lineWidth)), NSNumber(value: Double(space + lineWidth))]
        }
        layer.addSublayer(shapeLayer)
    }
}

@objc
extension UITextField {
    @objc(xt_placeholder:placeholderColor:)
    func xt_placeholder(_ placeholder: String, placeholderColor: UIColor) {
        attributedPlaceholder = NSAttributedString(
            string: placeholder,
            attributes: [.foregroundColor: placeholderColor]
        )
    }
}

@objc
extension UIViewController {
    @objc(xt_presentViewController:animated:completion:modalPresentationStyle:)
    func xt_presentViewController(
        _ viewControllerToPresent: UIViewController,
        animated flag: Bool,
        completion: (() -> Void)?,
        modalPresentationStyle: UIModalPresentationStyle
    ) {
        viewControllerToPresent.modalPresentationStyle = modalPresentationStyle
        present(viewControllerToPresent, animated: flag, completion: completion)
    }
}
